### PR TITLE
Get rid of horton.matrix

### PR DESCRIPTION
**Pull request by tovrstra**
Main changes, other than deleting horton.matrix:
- Orbital class in horton.meanfield.orbitals.
- exp* renamed to orb*, e.g. exp_alpha becomes orb_alpha.
- In most cases, Cholesky-decomposed operators require.
  different functions.
- Examples and documentation updated as well.
- More consistent API in GOBasis class.
- Simplified type checking in GOBasis class.
- Using memoryviews in GOBasis class.
- The rest of HORTON is made compatible with these changes.

We should probably postpone the merge until we have released 2.1.0

### CODE REVIEW

**Comment by tovrstra** [comment]
Fixes #239

**Comment by tovrstra** [comment]
Ready! All should be fine but please wait with merging until we have 2.1.0 out.

File: horton/modelhamiltonians/test/test_hubbard.py [hunk 1]
```diff
@@ -27,25 +27,24 @@ def test_1d_hubbard_hamiltonian():
...
         RDirectTerm(er, 'hartree'),
         RExchangeTerm(er, 'x_hf'),
     ]
     ham = REffHam(terms)
     scf_solver = PlainSCFSolver()
-    scf_solver(ham, lf, olp, occ_model, exp_alpha)
+    scf_solver(ham, olp, occ_model, orb_alpha)
     energy = ham.compute_energy()
 
     assert (abs(energy --7.94427) < 1e-4)
```
Comment (matt-chan) [inline horton/modelhamiltonians/test/test_hubbard.py]: Weird... Does that work?

File: horton/modelhamiltonians/test/test_hubbard.py [hunk 2]
```diff
@@ -55,25 +54,24 @@ def test_1d_hubbard_hamiltonian_no_pbc():
...
         RDirectTerm(er, 'hartree'),
         RExchangeTerm(er, 'x_hf'),
     ]
     ham = REffHam(terms)
     scf_solver = PlainSCFSolver()
-    scf_solver(ham, lf, olp, occ_model, exp_alpha)
+    scf_solver(ham, olp, occ_model, orb_alpha)
     energy = ham.compute_energy()
 
     assert (abs(energy --7.0533) < 1e-4)
```
Comment (matt-chan) [inline horton/modelhamiltonians/test/test_hubbard.py]: Same as above...

**Comment by matt-chan** [comment]
The Cholesky code isn't used very much I think. The style will need to change anyways I think because if we start adding things like DF/Kalay integrals etc, then the code will have too many if/else statements. We can always go back and amend it later.

File: data/examples/hf_compare/compare.py [hunk 3]
```diff
@@ -32,25 +32,25 @@ def main(fns_fchk):
...
+            orbs.append(mol.orb_beta)
 
         # Keep the g09 dms.
         if debug:
-            dms_g09 = [exp.to_dm() for exp in exps]
+            dms_g09 = [orb.to_dm() for orb in orbs]
 
         # Construct an initial guess.
-        guess_core_hamiltonian(olp, kin, na, *exps)
+        guess_core_hamiltonian(olp, kin + na, *orbs)
```
Comment (tovrstra) [inline data/examples/hf_compare/compare.py]: My intention was to simplify the API and also to have less type checking in the functions. If you look at the argument list of the old version and the documentation, it took too much Python skill to understand how it works.

File: horton/io/iodata.py [hunk 4]
```diff
@@ -237,6 +232,12 @@ def __init__(self, **kwargs):
         for key, value in kwargs.iteritems():
             setattr(self, key, value)
 
+    # Numpy array attributes that may require orbital basis reordering or sign correction.
```
Comment (tovrstra) [inline horton/io/iodata.py]: This is a temporary hack and it seemed too early to work out a nice solution. I'll ad some comments on how this should be done better when factoring the io code.

File: horton/modelhamiltonians/test/test_hubbard.py (hunk 1 above)
Comment (tovrstra) [inline horton/modelhamiltonians/test/test_hubbard.py]: It does but it is not nice. I'll make it more readable.

File: horton/modelhamiltonians/test/test_hubbard.py (hunk 2 above)
Comment (tovrstra) [inline horton/modelhamiltonians/test/test_hubbard.py]: Agreed. by the way, not my code originally...

**Comment by tovrstra** [comment]
I'm not sure yet what the best way would be to support all different ways of dealing with four-center integrals. I'd like to postpone such decisions, if that's ok.

**Pull request merged**